design-proposals/klip-2-produce-data.md

```diff
@@ -0,0 +1,347 @@
+without pre-existing data.
+
+## Public APIS
+
+#### INSERT INTO ... VALUES
+
+```sql
+INSERT INTO <stream_name|table_name> [(column1, column2, ...)]
+  VALUES (value1, value2, ...)
+```
+
+The `INSERT VALUES` statement can be used to insert new records into a KSQL source (agnostic to the
+Stream/Table duality). If the columns are not present, it is assumed that the values are in the same
+order as the schema and contain every field. `ROWKEY` is always present as the first column, and is 
+automatically populated with semantics described in the section on `ROWKEY` and `ROWTIME` below.
+
+The value for `stream_name`/`table_name` must be a valid Stream/Table registered with a KSQL schema.
+The serialization format will be the same as the format specified in the `value_format` of the
+`CREATE` statement. 
+
+For the initial version, only primitive literal types will be supported (e.g. `INT`, `VARCHAR`, 
+etc...) but syntax is easily extensible to support struct, map and list (see _Complex Datatypes_
+section below).
+
+Future extensions: 
+* Support Structs/Complex Types
+* Support some `AT INTERVAL` syntax to generalize the functionality to generate full stream data.
+* Read from file to allow reproducible insertions
+* Headless mode support
+
+#### DELETE FROM
+
+```sql 
+DELETE FROM table_name WHERE ROWKEY = <rowkey_value>;
+```
+
+The `DELETE` statement can be used to insert tombstones into kafka topics that back tables. If the
+source is not a table, the statement will fail. For this KLIP, we will only support conditions
+of the form `ROWKEY = value`. While this may initially be confusing to users coming from a pure SQL
+background (as `DELETE FROM` supports arbitrary conditions), maintaining consistent documentation 
+and examples, along with adequate error messages, will alleviate the concern.
+
+#### CREATE STREAM/TABLE
+
+```sql
+CREATE STREAM stream_name ( { column_name data_type } [, ...] )
+  WITH ( property_name = expression [, ...] );
+```
+
+```sql
+CREATE TABLE table_name ( { column_name data_type } [, ...] )
+  WITH ( property_name = expression [, ...] );
+```
+
+The `WITH` clause for CS/CT statements will now accept two additional property_names: 
+`PARTITIONS`/`REPLICAS` If the topic *does not exist* and these values are specified, this command 
+will create a topic with a the set number of partitions and replicas before executing the command. 
+
+If the topic does not exist, and `PARTITIONS`/`REPLICAS` are not supplied, the command will fail
+with the following error message: 
+
+> Topic 'foo' does not exist. If you require an empty topic to be created for the stream/table, 
+> please re-run the statement providing the required 'PARTITION' count and, optional, 'REPLICAS' 
+> count in the WITH clause, e.g. 
+> 
+>     CREATE STREAM BAR (f0, f1) WITH (Kafka_topic='foo', PARTITIONS=4, REPLICAS=2);
+
+If the topic exists and has a different number of partitions/replicas, this command will fail with
```

Comment:
As previously noted, CSAS / CTAS:
- check the number of partitions in the existing topic and fail if the number differs from that requested. Where the requested count comes either from the CSAS / CTAS statement or via config.
 - check the number of replicas in the existing topic and fail if the number is less that the requested.  Where the requested count comes either from the CSAS / CTAS statement or via config.

We should ensure KSQL doesn't fail CS / CT statements if the replica count has increased. Personally, I don't think we should check replica counts. They are purely there for resilience and we should allow cluster admins to increase or decrease as they see fit.

If the topic already exists with a different partition count, it is not clear to me from this text if it makes any difference if the partition count is supplied explicitly by the user in the statement or comes from a default set in config.  It's also not clear to me if KSQL will only check partition counts when the statement is initially submitted or also when executing the statement read from the command topic, (i.e. during distribution to over servers _and_ during recovery).  Would a downstream query fail to start on a restart if the source topic's partition count changed?  Do we want it to?

Reply:
> We should ensure KSQL doesn't fail CS / CT statements if the replica count has increased. Personally, I don't think we should check replica counts. They are purely there for resilience and we should allow cluster admins to increase or decrease as they see fit.

Reply:
> Yes and I'm pretty sure yes. I've rarely seen a valid case in production where semantics stay the same when partitions change, but perhaps you can explain one to me :)

The downstream query certainly _should_ fail.  My question was more about does it currently fail or not?